x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexing/AsyncTwoPhaseIndexer.java

```diff
@@ -94,16 +95,21 @@ public synchronized IndexerState start() {
      * @return The new state for the indexer (STOPPED, STOPPING or ABORTING if the job was already aborted).
      */
     public synchronized IndexerState stop() {
+        AtomicBoolean wasStartedAndSetStopped = new AtomicBoolean(false);
         IndexerState currentState = state.updateAndGet(previousState -> {
             if (previousState == IndexerState.INDEXING) {
                 return IndexerState.STOPPING;
             } else if (previousState == IndexerState.STARTED) {
-                onStop();
+                wasStartedAndSetStopped.set(true);
```

Comment:
`onStop` is now persisting state so it must be called after the state has been updated.